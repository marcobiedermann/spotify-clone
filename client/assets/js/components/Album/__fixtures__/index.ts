const artist = {
  id: '',
  name: '',
};

const image = {
  height: 100,
  url: '',
  width: 100,
};

const item = {
  duration_ms: 100,
  id: '',
  name: '',
  track_number: 100,
};

const album = {
  artists: [artist],
  id: '',
  images: [image],
  name: '',
  tracks: {
    items: [item],
  },
};

export default album;
