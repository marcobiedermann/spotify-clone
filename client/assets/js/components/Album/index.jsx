import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import './style.css';

const Album = ({
  artists, id, images, name,
}) => (
  <figure className="album">
    {images[1] && (
      <Link to={`/albums/${id}`}>
        <img src={images[1].url} alt={name} />
      </Link>
    )}
    <figcaption>
      <h3>
        <Link to={`/albums/${id}`}>
          {name}
        </Link>
      </h3>
      {artists[0] && (
        <h4>
          <Link to={`/artists/${artists[0].id}`}>
            {artists[0].name}
          </Link>
        </h4>
      )}
    </figcaption>
  </figure>
);

Album.propTypes = {
  artists: PropTypes.arrayOf(PropTypes.shape()),
  id: PropTypes.string,
  images: PropTypes.arrayOf(PropTypes.shape()),
  name: PropTypes.string,
};

Album.defaultProps = {
  artists: [],
  id: '',
  images: [],
  name: '',
};

export default Album;
