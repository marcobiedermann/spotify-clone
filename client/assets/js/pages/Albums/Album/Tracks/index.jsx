import React from 'react';
import { Helmet } from 'react-helmet';

const TracksPage = () => (
  <div>
    <Helmet>
      <title>Tracks</title>
    </Helmet>
    Tracks
  </div>
);

export default TracksPage;
