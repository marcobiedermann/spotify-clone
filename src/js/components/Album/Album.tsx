import React from 'react';
import { Link } from 'react-router-dom';
import AlbumTracks from '../AlbumTracks';
import Image from '../Image';

const defaultImageSize = 100;

interface Artist {
  id?: string;
  name?: string;
}

interface Image {
  height: number | null;
  url: string;
  width: number | null;
}

interface Item {
  artists?: Artist[];
  duration_ms?: number;
  id?: string;
  name?: string;
  track_number?: number;
}

interface Tracks {
  items: Item[];
}

export interface AlbumProps {
  artists?: Artist[];
  id: string;
  images: Image[];
  name: string;
  tracks?: Tracks;
}

function Album(props: AlbumProps): JSX.Element {
  const { artists = [], id, images, name, tracks } = props;
  const artist = artists[0];
  const image = images[0];

  return (
    <div>
      <figure>
        {image && (
          <Link to={`/albums/${id}`}>
            <Image
              {...image}
              alt={name}
              width={image.width || defaultImageSize}
              height={image.height || defaultImageSize}
            />
          </Link>
        )}
        <figcaption>
          <h3>
            <Link to={`/albums/${id}`}>{name}</Link>
          </h3>
          <h4>
            <Link to={`/artists/${artist.id}`}>{artist.name}</Link>
          </h4>
        </figcaption>
      </figure>
      {tracks && <AlbumTracks items={tracks.items} />}
    </div>
  );
}

export default Album;
