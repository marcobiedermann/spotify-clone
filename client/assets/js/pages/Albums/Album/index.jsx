import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import {
  Route,
  Switch,
} from 'react-router-dom';
import { bindActionCreators } from 'redux';
import TracksPage from './Tracks';
import { fetchAlbum } from '../../../actions/albums';
import Album from '../../../components/Album';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';

export class AlbumPage extends Component {
  componentDidMount() {
    const {
      accessToken,
      fetchAlbum,
      match,
    } = this.props;

    fetchAlbum(accessToken, match.params.album_id);
  }

  render() {
    const {
      album,
      error,
      isLoading,
      match,
    } = this.props;

    return (
      <Switch>
        <Route
          path={`${match.url}/tracks`}
          component={TracksPage}
        />
        <Route
          path={match.url}
          component={() => {
            if (error) {
              return <Error>{error.message}</Error>;
            }

            if (isLoading) {
              return <Loader />;
            }

            return (
              <div className="album-page">
                <Helmet>
                  <title>{album.name}</title>
                </Helmet>
                <Album {...album} />
              </div>
            );
          }}
        />
      </Switch>
    );
  }
}

AlbumPage.propTypes = {
  accessToken: PropTypes.string,
  album: PropTypes.shape({
    name: PropTypes.string,
  }),
  error: PropTypes.shape({
    message: PropTypes.string,
  }),
  fetchAlbum: PropTypes.func,
  isLoading: PropTypes.bool,
  match: PropTypes.shape({
    params: PropTypes.shape({
      album_id: PropTypes.string,
    }),
    url: PropTypes.string,
  }),
};

AlbumPage.defaultProps = {
  accessToken: '',
  album: {
    name: '',
  },
  error: {
    message: '',
  },
  fetchAlbum: () => {},
  isLoading: false,
  match: {
    params: {
      album_id: '',
    },
    url: '',
  },
};

const mapStateToProps = state => ({
  ...state,
  album: state.albums.album,
  error: state.albums.error,
  isLoading: state.albums.isLoading,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    fetchAlbum,
  },
  dispatch,
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AlbumPage);
