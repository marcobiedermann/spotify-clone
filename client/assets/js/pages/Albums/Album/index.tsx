import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { RouteChildrenProps, useParams } from 'react-router-dom';
import useSWR from 'swr';
import Album from '../../../components/Album';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';

interface Params {
  albumId: string;
}

const AlbumPage: FC<RouteChildrenProps> = () => {
  const { albumId } = useParams<Params>();
  const { data, error } = useSWR(`/v1/albums/${albumId}`);

  if (error) {
    return <Error>{error.message}</Error>;
  }

  if (!data) {
    return <Loader />;
  }

  const { name } = data;

  return (
    <>
      <Helmet>
        <title>{name}</title>
      </Helmet>
      <Album {...data} />
    </>
  );
};

export default AlbumPage;
