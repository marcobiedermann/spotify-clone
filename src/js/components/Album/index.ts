import Album from './Album';

export default Album;
export * from './Album';
