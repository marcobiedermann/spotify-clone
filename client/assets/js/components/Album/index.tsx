import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import { ArtistProps } from '../Artist';
import Image, { ImageProps } from '../Image';
import Tracks, { TracksProps } from '../Tracks';

export interface AlbumProps {
  artists: ArtistProps[];
  id: string;
  images: ImageProps[];
  name: string;
  tracks: TracksProps;
}

const Album: FC<AlbumProps> = (props) => {
  const { artists, id, images, name, tracks } = props;

  return (
    <div>
      <figure>
        {images[1] && (
          <Link to={`/albums/${id}`}>
            <Image {...images[1]} alt={name} />
          </Link>
        )}
        <figcaption>
          <h3>
            <Link to={`/albums/${id}`}>{name}</Link>
          </h3>
          {artists[0] && (
            <h4>
              <Link to={`/artists/${artists[0].id}`}>{artists[0].name}</Link>
            </h4>
          )}
        </figcaption>
      </figure>
      {tracks && <Tracks {...tracks} />}
    </div>
  );
};

export default Album;
