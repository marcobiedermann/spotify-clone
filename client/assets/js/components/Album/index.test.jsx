import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Album from '.';

describe('Album component', () => {
  const album = {
    artists: [
      {
        id: '123456',
      },
    ],
    id: '123456',
    images: [{}, {}],
    name: 'Album Name',
    tracks: {
      items: [],
    },
  };

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Album {...album} />
        </MemoryRouter>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
