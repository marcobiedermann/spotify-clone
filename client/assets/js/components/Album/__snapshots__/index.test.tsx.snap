// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album component renders correctly 1`] = `
<div>
  <figure>
    <figcaption>
      <h3>
        <a
          href="/albums/"
          onClick={[Function]}
        >
          
        </a>
      </h3>
      <h4>
        <a
          href="/artists/"
          onClick={[Function]}
        >
          
        </a>
      </h4>
    </figcaption>
  </figure>
  <table>
    <tbody>
      <tr>
        <td>
          100
        </td>
        <td>
          
        </td>
        <td>
          00:00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
